src/net_processing.cpp

```diff
@@ -3969,43 +3999,39 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
     return;
 }
 
-bool PeerManagerImpl::MaybeDiscourageAndDisconnect(CNode& pnode)
+bool PeerManagerImpl::MaybeDiscourageAndDisconnect(CNode& pnode, Peer& peer)
```

Comment:
`MaybeDiscourageAndDisconnect` should continue needing a `CNode` in order to signal the disconnection, I think.

Reply:
> MaybeDiscourageAndDisconnect should continue needing a CNode in order to signal the disconnection, I think.

This is true. It's also currently needed for the connection type, which I don't think belongs in net_processing.

> Even though you could grab the PeerRef from the m_peer_map in here again, passing in a Peer is a bit less overhead?

We want to grab a `PeerRef` at the entry functions (`ProcessMessages` and `SendMessages` for the message handler thread), and then pass that reference down through the call stack. Grabbing `PeerRef` incurs overhead because `m_peer_mutex` needs to be locked and the shared ptr's refcount needs to be incremented/decremented, so we shouldn't do it unnecessarily.